content/en/blog/_posts/2024-08-19-pod-failure-policy-ga.md

```diff
@@ -0,0 +1,172 @@
+---
+layout: blog
+title: "Kubernetes 1.31: Pod failure policy goes GA"
+date: 2024-08-19
+slug: kubernetes-1-31-pod-failure-policy-stable
+author: >
+  Michał Woźniak (Google)
+  Shannon Kularathna (Google)
+---
+
+This post describes _Pod failure policy_, which graduates to stable in Kubernetes
+1.31, and how to use it in your Jobs.
+
+## About Pod failure policy
+
+When you run workloads on Kubernetes, Pods might fail for a variety of reasons.
+Ideally, workloads like Jobs should be able to ignore transient, retriable
+failures and continue running to completion.
+
+To allow for these transient failures, Kubernetes Jobs include the `backoffLimit`
+field, which lets you specify a number of Pod failures that you're willing to tolerate
+during Job execution. However, if you set a large value for the `backoffLimit` field
+and rely solely on this field, you might notice unnecessary increases in operating
+costs as Pods restart excessively until the backoffLimit is met.
+
+This becomes particularly problematic when running large-scale Jobs with
+thousands of long-running Pods across thousands of nodes.
+
+The Pod failure policy extends the backoff limit mechanism to help you reduce
+costs in the following ways:
+
+- Gives you control to fail the Job as soon as a non-retriable Pod failure occurs.
+- Allows you to ignore retriable errors without increasing the `backoffLimit` field.
+
+For example, you can use a Pod failure policy to run your workload on more affordable spot machines
+by ignoring Pod failures caused by
+[graceful node shutdown](/docs/concepts/cluster-administration/node-shutdown/#graceful-node-shutdown).
+
+The policy allows you to distinguish between retriable and non-retriable Pod
+failures based on container exit codes or Pod conditions in a failed Pod.
+
+## How it works
+
+You specify a Pod failure policy in the Job specification, represented as a list
+of rules.
+
+For each rule you define _match requirements_ based on one of the following properties:
+
+- Container exit codes: the `onExitCodes` property.
+- Pod conditions: the `onPodConditions` property.
+
+Additionally, for each rule, you specify one of the following actions to take
+when a Pod matches the rule:
+- `Ignore`: Do not count the failure towards the `backoffLimit` or `backoffLimitPerIndex`.
+- `FailJob`: Fail the entire Job and terminate all running Pods.
+- `FailIndex`: Fail the index corresponding to the failed Pod
+  [Backoff limit per index](/docs/concepts/workloads/controllers/job/#backoff-limit-per-index),
+- `Count`: Count the failure towards the `backoffLimit` or `backoffLimitPerIndex`.
+  This is the default behavior.
+
+When Pod failures occur in a running Job, Kubernetes matches the
```

Comment:
I know that this blog post is about policy, but it may worth mentioning that Pod restartPolicy must be Never. To highlight that the node is not making the restart decision.

Reply:
Thanks for the suggestion. This is indeed an important design limitation. It is mentioned on the [Pod Failure Policy page](https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-failure-policy): "if you want to use a .spec.podFailurePolicy field for a Job, you must also define that Job's pod template with .spec.restartPolicy set to Never.".

We could consider saying something like: 
``` 
When you specify the Pod failure policy you must also define the Job's Pod template with
`restartPolicy: Never`. This is required to ensure there are no race conditions between
Kubelet and Job controller when counting Pod failures" (at the bottom of the "How it works".
```

OTOH, I think this might be too detailed for the tone of the blog post. WDYT @atiratree @shannonxtreme ?

Reply:
I think it could be useful to have, especially since we do not have a full example with the Job/Pod object in the blog.